### PR TITLE
Bump to 0.23

### PR DESCRIPTION
**Pull request by kylebarron**
### Change list 

- Bump pyo3-async-runtimes version in Cargo.toml and documentation to 0.23
- Bump pyo3 dependency version to 0.23
- Switched to new `IntoPyObject` trait, with a couple compilation errors still that I'm not sure how to fix.
- Fixed lints to remove `_bound` suffix

### CODE REVIEW

**Comment by kylebarron** [comment]
I tried two slightly different approaches that both hit small snags. It seems like we need _some_ bound on the `Target` associated type of `IntoPyObject`, but we probably also want that to be quite flexible? So the latest two commits added a new generic with a bound of `AsRef<PyAny>`. And that seemed to help, just getting a single compilation error now that I don't know how to fix.

File: src/generic.rs [hunk 1]
```diff
@@ -1467,28 +1480,34 @@ where
...
         match self.tx.try_send(item.clone_ref(py)) {
-            Ok(_) => Ok(true.into_py(py)),
+            Ok(_) => Ok(true.into_pyobject(py)?.as_ref().unbind()),
             Err(e) => {
                 if e.is_full() {
                     let mut tx = self.tx.clone();
                     Python::with_gil(move |py| {
-                        Ok(
-                            future_into_py_with_locals::<R, _, PyObject>(py, locals, async move {
+                        Ok(future_into_py_with_locals::<R, _, PyObject, _>(
```
Comment (kylebarron) [inline src/generic.rs:1488]: Getting this one error that I still don't know how to fix

```
error[E0277]: the trait bound `pyo3::PyAny: std::convert::AsRef<pyo3::PyAny>` is not satisfied
    --> src/generic.rs:1488:73
     |
1488 |                         Ok(future_into_py_with_locals::<R, _, PyObject, _>(
     |                                                                         ^ the trait `std::convert::AsRef<pyo3::PyAny>` is not implemented for `pyo3::PyAny`
     |
note: required by a bound in `generic::future_into_py_with_locals`
    --> src/generic.rs:590:8
     |
581  | pub fn future_into_py_with_locals<R, F, T, P>(
     |        -------------------------- required by a bound in this function
...
590  |     P: AsRef<PyAny>,
     |        ^^^^^^^^^^^^ required by this bound in `future_into_py_with_locals`

```

**Comment by kylebarron** [comment]
I noticed https://github.com/PyO3/pyo3/pull/4708, which helped me fix compilation here without needing to add another generic.

File: src/async_std.rs [hunk 2]
```diff
@@ -279,7 +279,7 @@ pub fn future_into_py_with_locals<F, T>(
 ) -> PyResult<Bound<PyAny>>
 where
     F: Future<Output = PyResult<T>> + Send + 'static,
-    T: IntoPy<PyObject>,
+    T: for<'py> IntoPyObject<'py>,
```
Comment (Icxolu) [inline src/async_std.rs:282]: These bounds are too strict. You should introduce a `'py` lifetime on function and tie it to the `Python<'py>` token, as well as `IntoPyObject<'py>`. You only need higher ranked bounds of there is no `'py` input lifetime. (Same for a lot of the others)

File: src/lib.rs [hunk 3]
```diff
@@ -576,14 +576,14 @@ impl PyEnsureFuture {
     }
 }
 
-fn call_soon_threadsafe(
-    event_loop: &Bound<PyAny>,
+fn call_soon_threadsafe<'py>(
+    event_loop: &'py Bound<PyAny>,
```
Comment (Icxolu) [inline src/lib.rs:580]:
```suggestion
    event_loop: &Bound<'py, PyAny>,
```
You want to tie this to the Python lifetime of the `Bound`.

File: src/generic.rs [hunk 4]
```diff
@@ -686,10 +694,10 @@ impl<T> Cancellable<T> {
     }
 }
 
-impl<F, T> Future for Cancellable<F>
+impl<'py, F, T> Future for Cancellable<F>
 where
     F: Future<Output = PyResult<T>>,
-    T: IntoPy<PyObject>,
+    T: IntoPyObject<'py>,
```
Comment (Icxolu) [inline src/generic.rs:700]: This one if _probably_ the only one where you need the higher ranked bound, because there is no input to tie it to.

**Review by Icxolu — COMMENTED** [review]
Thanks for helping with the migration!

I have some comments on the `IntoPyObject` migration. The majority of the bounds introduced here are stricter then necessary. The higher ranked bound (`for<'py> ...`) means the bound must hold for _any_ arbitrary lifetime. But most of the time you already have a specific `'py` lifetime as an input via the `Python<'py>` token, a `Bound<'py, T>` or a `Borrowed<'_, 'py, T>`. In these case it is only necessary to restrict the bound to that specific lifetime instead of _all_ possible lifetimes.

The higher ranked bound is needed when you don't have any input that brings a `'py` lifetime. In these cases the `Python<'py>` token is usually created via `Python::with_gil` which _can_ have any arbitrary lifetime.

I hope this clears it a bit up and maybe solves the compile errors you were running into.
(PyO3/pyo3#4708 should help with boilerplate, but everything should be possible to implement in the mean time without it)

**Comment by kylebarron** [comment]
@Icxolu I tried to move to the `'py` lifetime defined on the function, but that gave an error of 
```
error[E0521]: borrowed data escapes outside of closure
    --> src/generic.rs:1033:49
     |
989  | pub fn local_future_into_py_with_locals<'py, R, F, T>(
     |                                         --- lifetime `'py` defined here
...
1022 |             Python::with_gil(move |py| {
     |                                    --
     |                                    |
     |                                    `py` is a reference that is only valid in the closure body
     |                                    has type `pyo3::Python<'1>`
...
1033 |                     result.and_then(|val| match val.into_pyobject(py) {
     |                                                 ^^^^^^^^^^^^^^^^^^^^^
     |                                                 |
     |                                                 `py` escapes the closure body here
     |                                                 argument requires that `'1` must outlive `'py`
```
like you said [here](https://github.com/PyO3/pyo3-async-runtimes/pull/21#discussion_r1846993716).

But then when I switch just those two functions in `generic.rs` that internally call `Python::with_gil` to use the higher-ranked lifetime, I get an error [here](https://github.com/kylebarron/pyo3-async-runtimes/blob/944cbc82e209a8f47e232e679883aca5fddf397b/src/generic.rs#L1199) saying

```
error[E0277]: `T` cannot be converted to a Python object
    --> src/generic.rs:1199:46
     |
1199 |     local_future_into_py_with_locals::<R, F, T>(py, get_current_locals::<R>(py)?, fut)
     |                                              ^ the trait `for<'py> pyo3::IntoPyObject<'py>` is not implemented for `T`
     |
     = note: `IntoPyObject` is automatically implemented by the `#[pyclass]` macro
     = note: if you do not wish to have a corresponding Python type, implement it manually
     = note: if you do not own `T` you can perform a manual conversion to one of the types in `pyo3::types::*`
note: required by a bound in `generic::local_future_into_py_with_locals`
    --> src/generic.rs:997:8
     |
989  | pub fn local_future_into_py_with_locals<R, F, T>(
     |        -------------------------------- required by a bound in this function
...
997  |     T: for<'py> IntoPyObject<'py>,
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `local_future_into_py_with_locals`
```

So it seems that the higher-ranked `for<'py>` "infects" the rest of the functions. That's why I originally had the higher-ranked lifetime to get it to compile.

**Comment by Icxolu** [comment]
Ooops, I guess I spoke to soon then, sorry about that 🙈. On closer look `pyo3-async-runtimes` is a bit unique in the regard that it  passes around `Python` tokes, while still needing to use `Python::with_gil` because it `spawn`s, potentially crossing thread boundaries. In that case the higher ranked bound is indeed needed. (Looking at the signatures alone it just looked wrong 🙃)

File: src/lib.rs (hunk 3 above)
Comment (davidhewitt) [reply src/lib.rs:580]: Agreed on this one 👍

**Review by davidhewitt — COMMENTED** [review]
Thanks, overall this seems to be going in the right direction; looks like the `pytests` directory needs a bit of love too.

**Comment by kylebarron** [comment]
Tests seem to be passing locally now

**Comment by davidhewitt** [comment]
I think `-Zprofile` setting has been removed from nightly now that we have `-C instrument-coverage`, let me see if I can fix...

**Comment by davidhewitt** [comment]
Looks like this is all green, thanks for the help!

**Pull request merged**